docs/2019-09-09-Combining-Rewrite-Axioms.md

```diff
@@ -0,0 +1,200 @@
+
+
+In the
+[Polkadot Verification](https://github.com/runtimeverification/polkadot-verification/issues/20)
+issue, @ehildenb proposed that we should either apply a chain of rules to a
+symbolic configuration consisting of a single variable, or we should be able to
+combine rules into a single one. This document details the matching logic
+reasoning and assumptions behind these.
+
+Summary
+=======
+
+If we have this chain of axioms: `α₁(X₁)⇒β₁(X₁)` … `αₙ(Xₙ)⇒βₙ(Xₙ)` in which
+all the LHS are *function-like* and *injection-based*, and all the RHS
+except `βₙ(Xₙ)` are *function-like*, then their combined transition is
+
+```
+(⌈β₁(X₁)∧α₂(X₂)⌉ ∧ ⌈β₂(X₂)∧α₃(X₃)⌉ ∧ … ∧ ⌈βₙ₋₁(Xₙ₋₁)∧αₙ(Xₙ)⌉ ∧ α₁(X₁))→••…•βₙ(Xₙ)
+```
+
+TODO: We may need to figure out how to handle maps, sets, and other structures
+based on non-injective symbols.
+
+The general transformation
+==========================
+
+Let us take a configuration `φ(X)` which is functional,
+and which can be a single variable.
+
+Let us also take a configuration `α(X)→•β(X)`, with `α(X)` being functional.
+Then, similar to
+[basic symbolic execution algorithm](2018-11-08-Applying-Axioms.md), we have:
+
+```
+1. α(Y) → •β(Y) // the axiom
+1. α(Y) ∧ ⌈α(Y) ∧ φ(X)⌉ → (•β(Y)) ∧ ⌈α(Y) ∧ φ(X)⌉
+   // from (1) and propositional reasoning
+1. α(Y) ∧ ⌈α(Y) ∧ φ(X)⌉ = α(Y) ∧ φ(X)
+   // ML paper Prop. 5.24
+1. α(Y) ∧ φ(X) → (•β(Y)) ∧ ⌈α(Y) ∧ φ(X)⌉
+   // from (2) and (3)
+1. φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉ → (•β(Y)) ∧ ⌈α(Y) ∧ φ(X)⌉
+   // from (3) and (4)
+1. φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉ → ∃ Y . ((•β(Y)) ∧ ⌈α(Y) ∧ φ(X)⌉)
+   // FOL reasoning
+1. ∀ Y . (φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉ → ∃ Y . ((•β(Y)) ∧ ⌈α(Y) ∧ φ(X)⌉))
+   // All unquantified variables are universal
+1. ∀ Y . (φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉ → ∃ Y’ . ((•β(Y’)) ∧ ⌈α(Y’) ∧ φ(X)⌉))
+   // Renaming variables to make things clear
+1. ∀ Y .
+        ( φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉
+        → ∃ Y’ . ((•β(Y’)) ∧ ⌈α(Y’) ∧ φ(X)⌉ ∧ ⌈α(Y) ∧ φ(X)⌉)
+        )
+   // a ∧ b → c is the same as a ∧ b → c ∧ b
+```
+
+When doing normal rewriting, we usually expect to get substitutions for the
+variables in `Y` and `Y’` when computing `⌈α(Y) ∧ φ(X)⌉`, which usually allows
+us to remove these variables. However, when combining rewriting axioms,
+we don’t always get such substitutions, so we need to take a different approach.
+
+Note that the above also work for all patterns `φ` and `α`, not only functional
```

Comment:
From line 38, in order to apply prop 5.24 you need some functionality constraints; and because you apply it both ways, probably both  `φ` and `α` need to be functional.